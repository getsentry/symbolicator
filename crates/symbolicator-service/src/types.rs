//! Types for the Symbolicator API.
//!
//! This module contains some types which (de)serialise to/from JSON to make up the public
//! HTTP API.  Its messy and things probably need a better place and different way to signal
//! they are part of the public API.

use std::collections::BTreeMap;
use std::fmt;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use symbolicator_sources::ObjectType;

use crate::utils::hex::HexValue;

/// The scope of a source or debug file.
///
/// Based on scopes, access to debug files that have been cached is determined. If a file comes from
/// a public source, it can be used for any symbolication request. Otherwise, the symbolication
/// request must match the scope of a file.
#[derive(Debug, Clone, Deserialize, Serialize, Eq, Ord, PartialEq, PartialOrd, Hash)]
#[serde(untagged)]
#[derive(Default)]
pub enum Scope {
    #[serde(rename = "global")]
    #[default]
    Global,
    Scoped(Arc<str>),
}

impl AsRef<str> for Scope {
    fn as_ref(&self) -> &str {
        match *self {
            Scope::Global => "global",
            Scope::Scoped(ref s) => s,
        }
    }
}

impl fmt::Display for Scope {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Scope::Global => f.write_str("global"),
            Scope::Scoped(ref scope) => f.write_str(scope),
        }
    }
}

/// Configuration for scraping of JS Sources, Source Maps and Source Context from the web.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ScrapingConfig {
    /// Whether scraping should happen at all.
    pub enabled: bool,
    /// Whether Symbolicator should verify SSL certs when scraping from the web.
    ///
    /// Defaults to `true`, just to be safe.
    #[serde(default = "default_verify_ssl")]
    pub verify_ssl: bool,
    /// A list of "allowed origin patterns" that control:
    /// - for sourcemaps: what URLs we are allowed to scrape from.
    /// - for source context: which URLs should be authenticated using attached headers
    ///
    /// Allowed origins may be defined in several ways:
    /// - `http://domain.com[:port]`: Exact match for base URI (must include port).
    /// - `*`: Allow any domain.
    /// - `*.domain.com`: Matches domain.com and all subdomains, on any port.
    /// - `domain.com`: Matches domain.com on any port.
    /// - `*:port`: Wildcard on hostname, but explicit match on port.
    pub allowed_origins: Vec<String>,
    /// A map of headers to send with every HTTP request while scraping.
    pub headers: BTreeMap<String, String>,
}

impl Default for ScrapingConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            // verify_ssl: false,
            allowed_origins: vec!["*".to_string()],
            headers: Default::default(),
            verify_ssl: true,
        }
    }
}

fn default_verify_ssl() -> bool {
    true
}

/// Specification of a module loaded into the process.
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct RawObjectInfo {
    /// Platform image file type (container format).
    #[serde(rename = "type")]
    pub ty: ObjectType,

    /// Identifier of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_id: Option<String>,

    /// Name of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_file: Option<String>,

    /// Identifier of the debug file.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub debug_id: Option<String>,

    /// Name of the debug file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_file: Option<String>,

    /// Checksum of the file's contents.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_checksum: Option<String>,

    /// Absolute address at which the image was mounted into virtual memory.
    ///
    /// We do allow the `image_addr` to be skipped if it is zero. This is because systems like WASM
    /// do not require modules to be mounted at a specific absolute address. Per definition, a
    /// module mounted at `0` does not support absolute addressing.
    #[serde(default)]
    pub image_addr: HexValue,

    /// Size of the image in virtual memory.
    ///
    /// The size is infered from the module list if not specified.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub image_size: Option<u64>,
}

/// Information on a debug information file.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ObjectFileStatus {
    /// The file was found and successfully processed.
    Found,
    /// The image was not referenced in the stack trace and not further handled.
    #[default]
    Unused,
    /// The file could not be found in any of the specified sources.
    Missing,
    /// The file failed to process.
    Malformed,
    /// The file could not be downloaded.
    FetchingFailed,
    /// Downloading or processing the file took too long.
    Timeout,
    /// An internal error while handling this image.
    Other,
}

impl ObjectFileStatus {
    pub fn name(self) -> &'static str {
        // used for metrics
        match self {
            ObjectFileStatus::Found => "found",
            ObjectFileStatus::Unused => "unused",
            ObjectFileStatus::Missing => "missing",
            ObjectFileStatus::Malformed => "malformed",
            ObjectFileStatus::FetchingFailed => "fetching_failed",
            ObjectFileStatus::Timeout => "timeout",
            ObjectFileStatus::Other => "other",
        }
    }
}
