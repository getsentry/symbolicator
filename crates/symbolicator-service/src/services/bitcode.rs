//! Service to retrieve Apple Bitcode Symbol Maps.
//!
//! This service downloads and caches the `PList` and [`BcSymbolMap`] used to un-obfuscate
//! debug symbols for obfuscated Apple bitcode builds.

use std::fmt::{self, Display};
use std::sync::Arc;

use anyhow::Context;
use futures::future::{self, BoxFuture};
use sentry::{Hub, SentryFutureExt};

use symbolic::common::{ByteView, DebugId};
use symbolic::debuginfo::macho::{BcSymbolMap, UuidMapping};
use symbolicator_sources::{FileType, RemoteFile, SourceConfig};
use tempfile::NamedTempFile;

use crate::caching::{
    Cache, CacheEntry, CacheError, CacheItemRequest, CacheKey, CacheVersions, Cacher,
    SharedCacheRef,
};
use crate::services::download::DownloadService;
use crate::types::Scope;

use super::caches::versions::BITCODE_CACHE_VERSIONS;
use super::fetch_file;

/// Handle to a valid BCSymbolMap.
///
/// While this handle points to the raw data, this data is guaranteed to be valid, you can
/// only have this handle if a positive cache existed.
#[derive(Debug, Clone)]
pub struct BcSymbolMapHandle {
    pub file: RemoteFile,
    pub uuid: DebugId,
    pub data: ByteView<'static>,
}

impl BcSymbolMapHandle {
    /// Parses the map from the handle.
    pub fn bc_symbol_map(&self) -> anyhow::Result<BcSymbolMap<'_>> {
        BcSymbolMap::parse(&self.data).context("Failed to parse BCSymbolMap")
    }
}

/// The handle to be returned by [`CacheItemRequest`].
///
/// This trait requires us to return a handle regardless of its cache status.
/// This is this handle but we do not expose it outside of this module, see
/// [`BcSymbolMapHandle`] for that.
#[derive(Debug, Clone)]
struct CacheHandle {
    file: RemoteFile,
    uuid: DebugId,
    data: ByteView<'static>,
}

/// The kinds of DIFs which are stored in the cache accessed from [`FetchFileRequest`].
#[derive(Debug, Clone, Copy)]
enum AuxDifKind {
    BcSymbolMap,
    UuidMap,
}

impl Display for AuxDifKind {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            AuxDifKind::BcSymbolMap => write!(f, "BCSymbolMap"),
            AuxDifKind::UuidMap => write!(f, "UuidMap"),
        }
    }
}

/// The interface to the [`Cacher`] service.
///
/// The main work is done by the [`CacheItemRequest`] impl.
#[derive(Debug, Clone)]
struct FetchFileRequest {
    file_source: RemoteFile,
    uuid: DebugId,
    kind: AuxDifKind,
    download_svc: Arc<DownloadService>,
}

impl FetchFileRequest {
    #[tracing::instrument(skip(self, temp_file), fields(kind = %self.kind))]
    async fn fetch_auxdif(&self, temp_file: &mut NamedTempFile) -> CacheEntry {
        fetch_file(
            self.download_svc.clone(),
            self.file_source.clone(),
            temp_file,
        )
        .await?;

        let view = ByteView::map_file_ref(temp_file.as_file())?;

        match self.kind {
            AuxDifKind::BcSymbolMap => {
                if let Err(err) = BcSymbolMap::parse(&view) {
                    let kind = self.kind.to_string();
                    metric!(counter("services.bitcode.loaderrror") += 1, "kind" => &kind);
                    tracing::debug!("Failed to parse bcsymbolmap: {}", err);
                    return Err(CacheError::Malformed(err.to_string()));
                }
            }
            AuxDifKind::UuidMap => {
                if let Err(err) = UuidMapping::parse_plist(self.uuid, &view) {
                    let kind = self.kind.to_string();
                    metric!(counter("services.bitcode.loaderrror") += 1, "kind" => &kind);
                    tracing::debug!("Failed to parse plist: {}", err);
                    return Err(CacheError::Malformed(err.to_string()));
                }
            }
        }
        Ok(())
    }
}

impl CacheItemRequest for FetchFileRequest {
    type Item = Arc<CacheHandle>;

    const VERSIONS: CacheVersions = BITCODE_CACHE_VERSIONS;

    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheEntry> {
        Box::pin(self.fetch_auxdif(temp_file))
    }

    fn load(&self, data: ByteView<'static>) -> CacheEntry<Self::Item> {
        Ok(Arc::new(CacheHandle {
            file: self.file_source.clone(),
            uuid: self.uuid,
            data,
        }))
    }

    fn use_shared_cache(&self) -> bool {
        self.file_source.worth_using_shared_cache()
    }
}

#[derive(Debug, Clone)]
pub struct BitcodeService {
    cache: Arc<Cacher<FetchFileRequest>>,
    download_svc: Arc<DownloadService>,
}

impl BitcodeService {
    pub fn new(
        difs_cache: Cache,
        shared_cache: SharedCacheRef,
        download_svc: Arc<DownloadService>,
    ) -> Self {
        Self {
            cache: Arc::new(Cacher::new(difs_cache, shared_cache)),
            download_svc,
        }
    }

    /// Returns a `BCSymbolMap` if one is found for the `uuid`.
    pub async fn fetch_bcsymbolmap(
        &self,
        uuid: DebugId,
        scope: Scope,
        sources: Arc<[SourceConfig]>,
    ) -> Option<BcSymbolMapHandle> {
        // First find the PList.
        let plist_handle = self
            .fetch_file_from_all_sources(uuid, AuxDifKind::UuidMap, scope.clone(), sources.clone())
            .await?;

        let uuid_mapping = UuidMapping::parse_plist(uuid, &plist_handle.data)
            .context("Failed to parse plist")
            .map_err(|err| {
                tracing::warn!("{}: {:?}", err, err.source());
                sentry::capture_error(&*err);
                err
            })
            .ok()?;

        // Next find the BCSymbolMap.
        let symbolmap_handle = self
            .fetch_file_from_all_sources(
                uuid_mapping.original_uuid(),
                AuxDifKind::BcSymbolMap,
                scope,
                sources,
            )
            .await?;

        Some(BcSymbolMapHandle {
            file: symbolmap_handle.file.clone(),
            uuid: symbolmap_handle.uuid,
            data: symbolmap_handle.data.clone(),
        })
    }

    async fn fetch_file_from_all_sources(
        &self,
        uuid: DebugId,
        dif_kind: AuxDifKind,
        scope: Scope,
        sources: Arc<[SourceConfig]>,
    ) -> Option<Arc<CacheHandle>> {
        let file_type = match dif_kind {
            AuxDifKind::BcSymbolMap => &[FileType::BcSymbolMap],
            AuxDifKind::UuidMap => &[FileType::UuidMap],
        };
        let files = self
            .download_svc
            .list_files(&sources, file_type, &uuid.into())
            .await;

        let fetch_jobs = files.into_iter().map(|file_source| {
            let scope = if file_source.is_public() {
                Scope::Global
            } else {
                scope.clone()
            };
            let hub = Hub::new_from_top(Hub::current());
            hub.configure_scope(|scope| {
                scope.set_tag("auxdif.debugid", uuid);
                scope.set_extra("auxdif.kind", dif_kind.to_string().into());
                scope.set_extra("auxdif.source", file_source.source_metric_key().into());
            });
            let request = FetchFileRequest {
                file_source,
                uuid,
                kind: dif_kind,
                download_svc: self.download_svc.clone(),
            };
            let cache_key = CacheKey::from_scoped_file(&scope, &request.file_source);
            self.cache
                .compute_memoized(request, cache_key)
                .bind_hub(hub)
        });

        let all_results = future::join_all(fetch_jobs).await;
        all_results.into_iter().find_map(Result::ok)
    }
}
