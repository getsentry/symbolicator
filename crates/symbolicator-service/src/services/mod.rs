//! Provides the internal shared Symbolicator services and a way to initialize them.
//!
//! Symbolicator operates a number of independent services defined in this module for downloading,
//! cache management, and file access.
//! [`SharedServices`] initializes all these internal services according to the provided [`Config`].
//!
//! The internal services require a separate asynchronous runtimes dedicated for I/O-intensive work,
//! such as downloads and access to the shared cache.

use std::sync::Arc;

use anyhow::{Context, Result};

use crate::caching::{Caches, SharedCacheRef, SharedCacheService};
use crate::config::Config;

pub mod bitcode;
pub mod caches;
pub mod cficaches;
pub mod derived;
pub mod download;
mod fetch_file;
pub mod il2cpp;
mod minidump;
mod module_lookup;
pub mod objects;
pub mod ppdb_caches;
pub mod symbolication;
pub mod symcaches;

use self::caches::SourceFilesCache;
use self::download::DownloadService;
use self::objects::ObjectsActor;

pub use self::symbolication::ScrapingConfig;
pub use fetch_file::fetch_file;

pub struct SharedServices {
    pub config: Config,
    pub caches: Caches,
    pub download_svc: Arc<DownloadService>,
    pub shared_cache: SharedCacheRef,
    pub objects: ObjectsActor,
    pub sourcefiles_cache: Arc<SourceFilesCache>,
}

impl SharedServices {
    pub fn new(config: Config, io_pool: tokio::runtime::Handle) -> Result<Self> {
        let caches = Caches::from_config(&config).context("failed to create local caches")?;
        caches
            .clear_tmp(&config)
            .context("failed to clear tmp caches")?;

        let download_svc = DownloadService::new(&config, io_pool.clone());

        let shared_cache = SharedCacheService::new(config.shared_cache.clone(), io_pool);

        let sourcefiles_cache = Arc::new(SourceFilesCache::new(
            caches.sourcefiles.clone(),
            shared_cache.clone(),
            download_svc.clone(),
        ));

        let objects = ObjectsActor::new(
            caches.object_meta.clone(),
            caches.objects.clone(),
            shared_cache.clone(),
            download_svc.clone(),
        );

        Ok(Self {
            config,
            caches,
            download_svc,
            shared_cache,
            objects,
            sourcefiles_cache,
        })
    }
}
