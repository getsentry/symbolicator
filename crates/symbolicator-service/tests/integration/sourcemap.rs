use std::str::FromStr;
use std::sync::Arc;

use reqwest::Url;
use serde_json::json;
use symbolicator_service::services::ScrapingConfig;
use symbolicator_service::types::{JsFrame, RawObjectInfo, Scope};
use symbolicator_service::{
    services::symbolication::SymbolicateJsStacktraces, types::JsStacktrace,
};
use symbolicator_sources::{SentrySourceConfig, SourceId};

use crate::{assert_snapshot, setup_service};

/*
 * NOTES:
 *
 * - `cargo test --package symbolicator-service --test integration sourcemap` to run just tests below
 * - all test names are 1:1 to those in monolith for easier tracking
 * - original stacktrace frames are used in top-most frame first order, which is reverse of monolith
 *   is using, thus when you copy test from there, just reverse the json payload
 * - when copying the json payload, make sure to translate all escaped characters and python keywords
 *
 * TODOS:
 *
 * - add all error assertions, by querying for `event.data["errors"]` in monolith
 */

#[track_caller]
fn make_js_request(
    source: SentrySourceConfig,
    frames: &str,
    modules: &str,
    release: impl Into<Option<String>>,
    dist: impl Into<Option<String>>,
) -> SymbolicateJsStacktraces {
    let frames: Vec<JsFrame> = serde_json::from_str(frames).unwrap();
    let modules: Vec<RawObjectInfo> = serde_json::from_str(modules).unwrap();
    let stacktraces = vec![JsStacktrace { frames }];

    SymbolicateJsStacktraces {
        scope: Scope::Global,
        source: Arc::new(source),
        stacktraces,
        modules,
        release: release.into(),
        dist: dist.into(),
        scraping: ScrapingConfig {
            enabled: false,
            ..Default::default()
        },
        apply_source_context: true,
    }
}

fn sourcemap_server<L>(
    fixtures_dir: &str,
    lookup: L,
) -> (symbolicator_test::Server, SentrySourceConfig)
where
    L: Fn(&str, &str) -> serde_json::Value + Clone + Send + 'static,
{
    let fixtures_dir = symbolicator_test::fixture(format!("sourcemaps/{fixtures_dir}"));
    symbolicator_test::sourcemap_server(fixtures_dir, lookup)
}

#[tokio::test]
async fn test_sourcemap_expansion() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("01_sourcemap_expansion", |url, _query| {
        json!([{
            "type": "file",
            "id": "1",
            "url": format!("{url}/test.min.js"),
            "abs_path": "~/test.min.js",
        }, {
            "type": "file",
            "id": "2",
            "url": format!("{url}/test.min.js.map"),
            "abs_path": "~/test.min.js.map",
        }])
    });

    let frames = r#"[{
        "abs_path": "http://example.com/index.html",
        "filename": "index.html",
        "lineno": 6,
        "colno": 7,
        "function": "produceStack"
    }, {
        "abs_path": "http://example.com/test.min.js",
        "filename": "test.min.js",
        "lineno": 1,
        "colno": 183,
        "function": "i"
    }, {
        "abs_path": "async http://example.com/test.min.js",
        "filename": "test.min.js",
        "lineno": 1,
        "colno": 136,
        "function": "r"
    }, {
        "abs_path": "http://example.com/test.min.js",
        "filename": "test.min.js",
        "lineno": 1,
        "colno": 64,
        "function": "e"
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_sourcemap_source_expansion() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("02_sourcemap_source_expansion", |url, _query| {
        json!([{
            "type": "file",
            "id": "1",
            "url": format!("{url}/file.min.js"),
            "abs_path": "~/file.min.js",
        }, {
            "type": "file",
            "id": "2",
            "url": format!("{url}/file.min.js.map"),
            "abs_path": "~/file.min.js.map",
        }, {
            "type": "file",
            "id": "3",
            "url": format!("{url}/file1.js"),
            "abs_path": "~/file1.js",
        }, {
            "type": "file",
            "id": "4",
            "url": format!("{url}/file2.js"),
            "abs_path": "~/file2.js",
        }])
    });

    let frames = r#"[{
        "function": "function: \"HTMLDocument.<anonymous>\"",
        "abs_path": "http://example.com/index.html",
        "filename": "index.html",
        "lineno": 283,
        "colno": 17,
        "in_app": false
    }, {
        "abs_path": "http://example.com/file.min.js",
        "filename": "file.min.js",
        "lineno": 1,
        "colno": 39
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_sourcemap_embedded_source_expansion() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) =
        sourcemap_server("03_sourcemap_embedded_source_expansion", |url, _query| {
            json!([{
                "type": "file",
                "id": "1",
                "url": format!("{url}/embedded.js"),
                "abs_path": "~/embedded.js",
            }, {
                "type": "file",
                "id": "2",
                "url": format!("{url}/embedded.js.map"),
                "abs_path": "~/embedded.js.map",
            }])
        });

    let frames = r#"[{
        "function": "function: \"HTMLDocument.<anonymous>\"",
        "abs_path": "http://example.com/index.html",
        "filename": "index.html",
        "lineno": 283,
        "colno": 17,
        "in_app": false
    }, {
        "abs_path": "http://example.com/embedded.js",
        "filename": "embedded.js",
        "lineno": 1,
        "colno": 39
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_source_expansion() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("04_source_expansion", |url, _query| {
        json!([{
            "type": "file",
            "id": "1",
            "url": format!("{url}/foo.js"),
            "abs_path": "~/foo.js",
        }])
    });

    let frames = r#"[{
        "abs_path": "http://example.com/foo.js",
        "filename": "foo.js",
        "lineno": 1,
        "colno": 0
    }, {
        "abs_path": "http://example.com/foo.js",
        "filename": "foo.js",
        "lineno": 4,
        "colno": 0
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_inlined_sources() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("05_inlined_sources", |url, _query| {
        json!([{
            "type": "file",
            "id": "1",
            "url": format!("{url}/test.min.js"),
            "abs_path": "~/test.min.js",
        }])
    });

    let frames = r#"[{
        "abs_path": "http://example.com/test.min.js",
        "filename": "test.js",
        "lineno": 1,
        "colno": 1
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_sourcemap_nofiles_source_expansion() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) =
        sourcemap_server("06_sourcemap_nofiles_source_expansion", |url, _query| {
            json!([{
                "type": "file",
                "id": "1",
                "url": format!("{url}/nofiles.js"),
                "abs_path": "~/nofiles.js",
            }, {
                "type": "file",
                "id": "2",
                "url": format!("{url}/nofiles.js.map"),
                "abs_path": "~/nofiles.js.map",
            }])
        });

    let frames = r#"[{
        "abs_path": "app:///nofiles.js",
        "lineno": 1,
        "colno": 39
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_indexed_sourcemap_source_expansion() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) =
        sourcemap_server("07_indexed_sourcemap_source_expansion", |url, _query| {
            json!([{
                "type": "file",
                "id": "1",
                "url": format!("{url}/indexed.min.js"),
                "abs_path": "~/indexed.min.js",
            }, {
                "type": "file",
                "id": "2",
                "url": format!("{url}/indexed.min.js.map"),
                "abs_path": "~/indexed.min.js.map",
            }, {
                "type": "file",
                "id": "3",
                "url": format!("{url}/file1.js"),
                "abs_path": "~/file1.js",
            }, {
                "type": "file",
                "id": "4",
                "url": format!("{url}/file2.js"),
                "abs_path": "~/file2.js",
            }])
        });

    let frames = r#"[{
        "abs_path": "http://example.com/indexed.min.js",
        "filename": "indexed.min.js",
        "lineno": 1,
        "colno": 39
    }, {
        "abs_path": "http://example.com/indexed.min.js",
        "filename": "indexed.min.js",
        "lineno": 2,
        "colno": 44
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_malformed_abs_path() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("", |_url, _query| json!([]));

    // Missing colon was removed on purpose.
    let frames = r#"[{
        "abs_path": "http//example.com/test.min.js",
        "filename": "test.js",
        "lineno": 1,
        "colno": 1
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_fetch_error() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (srv, source) = sourcemap_server("", |_url, _query| json!([]));

    let missing_asset_url_1 = srv.url("/assets/missing_foo.js");
    let missing_asset_url_2 = srv.url("/assets/missing_bar.js");

    let frames = format!(
        r#"[{{
            "abs_path": "{missing_asset_url_1}",
            "filename": "foo.js",
            "lineno": 1,
            "colno": 0
        }}, {{
            "abs_path": "{missing_asset_url_2}",
            "filename": "foo.js",
            "lineno": 4,
            "colno": 0
        }}]"#
    );

    let mut request = make_js_request(source, &frames, "[]", String::from("release"), None);
    request.scraping.enabled = true;
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_invalid_location() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("08_sourcemap_invalid_location", |url, _query| {
        json!([{
            "type": "file",
            "id": "1",
            "url": format!("{url}/invalidlocation.js"),
            "abs_path": "~/invalidlocation.js",
        }, {
            "type": "file",
            "id": "2",
            "url": format!("{url}/invalidlocation.js.map"),
            "abs_path": "~/invalidlocation.js.map",
        }])
    });

    let frames = r#"[{
        "abs_path": "http://example.com/invalidlocation.js",
        "filename": "invalidlocation.js",
        "lineno": 0,
        "colno": 0,
        "function": "e"
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_webpack() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("09_webpack", |url, _query| {
        json!([{
            "type": "file",
            "id": "1",
            "url": format!("{url}/test1.min.js"),
            "abs_path": "~/test1.min.js",
        }, {
            "type": "file",
            "id": "2",
            "url": format!("{url}/test1.min.js.map"),
            "abs_path": "~/test1.min.js.map",
        }, {
            "type": "file",
            "id": "3",
            "url": format!("{url}/test2.min.js"),
            "abs_path": "~/test2.min.js",
        }, {
            "type": "file",
            "id": "4",
            "url": format!("{url}/test2.min.js.map"),
            "abs_path": "~/test2.min.js.map",
        }])
    });

    let frames = r#"[{
        "abs_path": "http://example.com/test1.min.js",
        "filename": "test1.min.js",
        "lineno": 1,
        "colno": 183,
        "function": "i"
    }, {
        "abs_path": "http://example.com/test2.min.js",
        "filename": "test2.min.js",
        "lineno": 1,
        "colno": 183,
        "function": "i"
    }]"#;

    let request = make_js_request(source, frames, "[]", String::from("release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn e2e_node_debugid() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("e2e_node_debugid", |url, query| {
        assert_eq!(query, "debug_id=2f259f80-58b7-44cb-d7cd-de1505e7e718");
        json!([{
            "type": "bundle",
            "id": "1",
            "url": format!("{url}/bundle.zip"),
        }])
    });

    let frames = r#"[{
        "abs_path": "/Users/lucaforstner/code/github/getsentry/sentry-javascript-bundler-plugins/packages/playground/öut path/rollup/entrypoint1.js",
        "lineno": 73,
        "colno": 36,
        "function": "Object.<anonymous>"
    }]"#;
    let modules = r#"[{
        "code_file": "/Users/lucaforstner/code/github/getsentry/sentry-javascript-bundler-plugins/packages/playground/öut path/rollup/entrypoint1.js",
        "debug_id": "2f259f80-58b7-44cb-d7cd-de1505e7e718",
        "type": "sourcemap"
    }]"#;

    let request = make_js_request(source, frames, modules, None, None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn e2e_source_no_header() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("e2e_source_no_header", |url, _query| {
        // NOTE: this bundle was manually constructed and has a `"source"` file without a
        // `Sourcemap` header, but with a `sourceMappingURL`.
        // The test files are the same as in `01_sourcemap_expansion`, but with a different
        // base url.
        json!([{
            "type": "bundle",
            "id": "1",
            "url": format!("{url}/bundle.zip"),
        }])
    });

    let frames = r#"[{
        "abs_path": "app:///_next/server/pages/_error.js",
        "lineno": 1,
        "colno": 64,
        "function": "e"
    }]"#;
    let modules = r#"[]"#;

    let request = make_js_request(source, frames, modules, String::from("some-release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn e2e_react_native() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = sourcemap_server("e2e_react_native", |url, _query| {
        json!([{
            "type": "file",
            "id": "1",
            "url": format!("{url}/index.android.bundle"),
            "abs_path": "~/index.android.bundle",
            "headers": { "SoUrCemAp": "index.android.bundle.map" }
        }, {
            "type": "file",
            "id": "2",
            "url": format!("{url}/index.android.bundle.map"),
            "abs_path": "~/index.android.bundle.map",
            "headers": {}
        }])
    });

    let frames = r#"[{
        "abs_path": "app:///index.android.bundle",
        "lineno": 1,
        "colno": 11940
    }]"#;
    let modules = r#"[]"#;

    let request = make_js_request(source, frames, modules, String::from("some-release"), None);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn e2e_multiple_smref_scraped() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (srv, source) = sourcemap_server("e2e_multiple_smref_scraped", |_url, _query| json!([]));

    let url = srv.url("/files/");
    let frames = format!(
        r#"[{{
        "abs_path": "{url}app.js",
        "lineno": 1,
        "colno": 64
    }}]"#
    );
    let modules = r#"[]"#;

    let mut request = make_js_request(source, &frames, modules, None, None);
    request.scraping.enabled = true;

    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}

// A manually triggered test that can be used to locally debug monolith behavior. Requires a list
// of frames, modules, release/dist pair and valid token, which can all be obtained from the event's
// JSON payload. We can modify this util to pull the data from JSON API directly if we use it more often.
//
// Run with:
// RUST_LOG=debug cargo test --package 'symbolicator-service' --test 'integration' -- 'sourcemap::test_manual_processing' --include-ignored
#[tokio::test]
#[ignore]
async fn test_manual_processing() {
    let token = "token";
    let org = "org";
    let project = "project";
    let release = None;
    let dist = None;
    let frames = r#"[{
        "function": "x",
        "module": "x",
        "filename": "x",
        "abs_path": "x",
        "lineno": 1,
        "colno": 1,
        "in_app": true
    }]"#;
    let modules = r#"[{
        "code_file": "x",
        "debug_id": "x",
        "type": "x"
    }]"#;

    let (symbolication, _) = setup_service(|_| ());
    let source = SentrySourceConfig {
        id: SourceId::new("sentry:project"),
        url: Url::from_str(&format!(
            "https://sentry.io/api/0/projects/{org}/{project}/artifact-lookup/"
        ))
        .unwrap(),
        token: token.to_string(),
    };

    let request = make_js_request(source, frames, modules, release, dist);
    let response = symbolication.symbolicate_js(request).await;

    assert_snapshot!(response.unwrap());
}
