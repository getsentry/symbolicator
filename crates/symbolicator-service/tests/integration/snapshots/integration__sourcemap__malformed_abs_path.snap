---
source: crates/symbolicator-service/tests/integration/sourcemap.rs
expression: response.unwrap()
---
stacktraces:
  - frames:
      - status: missing_sourcemap
        filename: test.js
        abs_path: http//example.com/test.min.js
        lineno: 1
        colno: 1
raw_stacktraces:
  - frames:
      - filename: test.js
        abs_path: http//example.com/test.min.js
        lineno: 1
        colno: 1

